pkg/api/types.go

```diff
@@ -180,6 +180,112 @@ type VolumeSource struct {
 	GitRepo *GitRepo `json:"gitRepo"`
 	// Secret represents a secret that should populate this volume.
 	Secret *SecretSource `json:"secret"`
+
+	// TODO (per thockin) https://github.com/GoogleCloudPlatform/kubernetes/pull/4055#discussion_r24865963
+	AWSElasticBlockStore *AWSElasticBlockStore `json:"elasticBlockStore"`
+
+	NFSMount *NFSMount `json:"nfsMount"`
+
+	PersistentVolumeClaimAttachment *PersistentVolumeClaimAttachment `json:"persistentVolumeClaim,omitempty"`
+}
+
+type PersistentVolume struct {
+	TypeMeta   `json:",inline"`
+	ObjectMeta `json:"metadata,omitempty"`
+
+	//Spec defines a persistent volume owned by the cluster
+	Spec PersistentVolumeSpec `json:"spec,omitempty"`
+
+	// Status represents the current information about persistent volume.
+	Status PersistentVolumeStatus `json:"status,omitempty"`
+}
+
+type PersistentVolumeSpec struct {
+	Capacity ResourceList `json:"capacity,omitempty`
+	// AccessModeTypes are inferred from the Source
+	Source VolumeSource `json:",inline"`
```

Comment:
I also have concerns about using VolumeSource here. Maybe create another object that splits out the remote volumes and inline it into VolumeSource.